types/index.d.ts

```diff
@@ -0,0 +1,228 @@
+     *
+     * @returns The processor on which freeze is invoked.
+     */
+    freeze(): Processor
+  }
+
+  type Plugin = Attacher
+  type Settings = {
+    [key: string]: unknown
+  }
+  /**
+   * Presets provide a potentially sharable way to configure processors.
+   * They can contain multiple plugins and optionally settings as well.
+   */
+  interface Preset {
+    plugins?: PluggableList
+    settings?: Settings
+  }
+  type PluginTuple = [Plugin, Settings]
+  type Pluggable = Plugin | Preset | PluginTuple
+  type PluggableList = Pluggable[]
+
+  /**
+   * An attacher is the thing passed to `use`.
+   * It configures the processor and in turn can receive options.
+   *
+   * Attachers can configure processors, such as by interacting with parsers and compilers, linking them to other processors, or by specifying how the syntax tree is handled.
+   *
+   * @this Processor context object is set to the invoked on processor.
+   * @param options Configuration
+   * @returns Optional.
+   */
+  interface Attacher {
+    (this: Processor, options?: unknown): Transformer | void
+  }
+
+  /**
+   * Transformers modify the syntax tree or metadata of a file. A transformer is a function which is invoked each time a file is passed through the transform phase.
+   * If an error occurs (either because it’s thrown, returned, rejected, or passed to `next`), the process stops.
+   *
+   * The transformation process in unified is handled by `trough`, see it’s documentation for the exact semantics of transformers.
+   *
+   * @param node
+   * @param file
+   * @param next If the signature of a transformer includes `next` (third argument), the function may finish asynchronous, and must invoke `next()`.
+   * @returns
+   * - `Error` — Can be returned to stop the process
+   * - `Node` — Can be returned and results in further transformations and `stringify`s to be performed on the new tree
+   * - `Promise` — If a promise is returned, the function is asynchronous, and must be resolved (optionally with a `Node`) or rejected (optionally with an `Error`)
+   */
+  interface Transformer {
+    (
+      node: Unist.Node,
+      file: VFileCompatible,
+      next?: (error: Error | null, tree: Unist.Node, file: vfile.VFile) => {}
+    ): Error | Unist.Node | Promise<Unist.Node>
+  }
+
+  class Parser {
+    parse(file: VFileCompatible): Unist.Node
+  }
+
+  type ParserFunction = (file: VFileCompatible) => Unist.Node
+
+  class Compiler {
+    compile(node: Unist.Node, file?: VFileCompatible): string
+  }
+  type CompilerFunction = (node: Unist.Node, file?: VFileCompatible) => string
+
+  type RunCallback = (
+    error: Error | null,
+    node: Unist.Node,
+    file: vfile.VFile
+  ) => void
+
+  type ProcessCallback = (error: Error | null, file: vfile.VFile) => void
+
+  type VFileCompatible = vfile.VFile | vfile.VFileOptions | string | Buffer
```

Comment:
```suggestion
  type VFileCompatible = vfile.VFile | vfile.VFileOptions | vfile.VFileContents
```

I think this works as well?

Reply:
LGTM! I'll replace it!

Reply:
Btw, I think we should define `VFileCompatible` to `vfile` later. Let's do it later when we move its types from DefinitelyTyped repo.